src/linker/Linker.Dataflow/DynamicallyAccessedMembersTypeHierarchy.cs

```diff
@@ -0,0 +1,197 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Mono.Cecil;
+using Mono.Linker.Steps;
+
+namespace Mono.Linker.Dataflow
+{
+	class DynamicallyAccessedMembersTypeHierarchy
+	{
+		readonly LinkContext _context;
+		readonly MarkStep _markStep;
+
+		// Cache of DynamicallyAccessedMembers annotations applied to types and their hierarchies
+		// Values
+		//   annotation - the aggregated annotation value from the entire base and interface hierarchy of the given type
+		//                If the type has a base class with annotation a1 and an interface with annotation a2, the stored
+		//                annotation is a1 | a2.
+		//   applied - set to true once the annotation was applied to the type
+		//             This only happens once the right reflection pattern is found.
+		//             If a new type is being marked and one of its base types/interface has the applied set to true
+		//             the new type will apply its annotation and will also set its applied to true.
+		// Non-interface types
+		//   - Only marked types with non-empty annotation are put into the cache
+		//   - Non-marked types are not stored in the cache
+		//   - Marked types which are not in the cache don't have any annotation
+		// Interface types
+		//   - All interface types accessible from marked types are stored in the cache
+		//   - If the interface type doesn't have annotation the value None is stored here
+		//
+		// It's not possible to use the marking as a filter for interfaces in the cache
+		// because interfaces are marked late and in effectively random order.
+		// For this cache to be effective we need to be able to fill it for all base types and interfaces
+		// of a type which is currently being marked - at which point the interfaces are not yet marked.
+		readonly Dictionary<TypeDefinition, (DynamicallyAccessedMemberTypes annotation, bool applied)> _typesInDynamicallyAccessedMembersHierarchy;
```

Comment:
This cache will include all interfaces reachable from marked types - which can be relatively large number (1000s). It is built regardless of the annotations being used or even present, since it acts as a negative cache as well.

Alternatively this could be implemented by re-computing the annotation for interface each type it's needed - basically trade memory or CPU. I went with the cache as some kind of cache would still be needed for the `applied` bit, and the memory impact doesn't seem that bad (couple of KB).